wagtail/wagtail
# Allow manual lock even if `WorkflowLock` is currently applied

Fix #9943.

Other than the lock descriptions, the main difference here is that the "Lock" button is shown (provided the user has the permission to do so) when a page/snippet is in a workflow. Previously, the button has been missing since Wagtail 4.0.

<img width="461" alt="image" src="https://user-images.githubusercontent.com/6379424/221629744-f1b25080-178a-4b58-a5c6-0a6506bf7dab.png">

How to test:

- Spin up bakerydemo
- Sign in as an editor
- Submit a page/snippet to a workflow
  - For snippets, configure it first. One way to do this is by making the `Person` snippet extend both `LockableMixin` and `WorkflowMixin` (see [docs](https://docs.wagtail.org/en/stable/topics/snippets.html#enabling-workflows-for-snippets)), running the migrations, then assigning the default "Moderators approval" workflow to the `Person` model.
- The status side panel should say "Locked by workflow", with no option to unlock it
- Sign in with a user account that has an edit and lock permission on the page or snippet, e.g. a moderator
- The status side panel should say "Unlocked" with an option to lock it. The description should say `"Reviewers can edit this %(model_name)s, lock it to prevent other reviewers from editing"`
- Once locked, for users with review permissions on the task, the behaviour should be the same as if the page was locked with the manual lock outside of a workflow. This means that other reviewers cannot edit the page. If they want to, they need to unlock it first (if they have the permissions to do so)

## Notes

- Our designers have broken down the main possible scenarios on [Figma](https://www.figma.com/file/h67EsVXdWsfu38WGGxWfpi/Wagtail-CMS?node-id=10861%3A133304&t=eB2SttDD8H25QBtU-1)
- The messages in the side panel mostly follow the designs, but I've taken a few liberties:
  - Locked by other user -> Locked by another user
  - Removing periods from the lock description to be more consistent with other status side panel items
- Some design updates, notably the use of the switch toggle component and changes to the message banner colours, will be done in a separate PR.
  - #10171



_Please check the following:_

    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)
